bootstrap/src/main/java/com/splunk/opentelemetry/javaagent/bootstrap/metrics/jmx/JmxMetricsWatcher.java

```diff
@@ -0,0 +1,193 @@
+          asList(
+              MBeanServerNotification.REGISTRATION_NOTIFICATION,
+              MBeanServerNotification.UNREGISTRATION_NOTIFICATION));
+
+  private final MBeanServer mBeanServer;
+  private final ExecutorService executorService;
+  private final MeterRegistry meterRegistry;
+  private final JmxQuery query;
+  private final MetersFactory metersFactory;
+
+  private final Map<ObjectName, List<Meter>> meters = new ConcurrentHashMap<>();
+  private volatile NotificationListener notificationListener;
+
+  public JmxMetricsWatcher(JmxQuery query, MetersFactory metersFactory) {
+    this(
+        findMBeanServer(),
+        Executors.newSingleThreadExecutor(daemonThreadFactory),
+        Metrics.globalRegistry,
+        query,
+        metersFactory);
+  }
+
+  // visible for tests
+  JmxMetricsWatcher(
+      MBeanServer mBeanServer,
+      ExecutorService executorService,
+      MeterRegistry meterRegistry,
+      JmxQuery query,
+      MetersFactory metersFactory) {
+    this.mBeanServer = mBeanServer;
+    this.executorService = executorService;
+    this.meterRegistry = meterRegistry;
+    this.query = query;
+    this.metersFactory = metersFactory;
+  }
+
+  // copied from micrometer CommonsObjectPool2Metrics class
+  private static MBeanServer findMBeanServer() {
+    List<MBeanServer> mBeanServers = MBeanServerFactory.findMBeanServer(null);
+    if (!mBeanServers.isEmpty()) {
+      return mBeanServers.get(0);
+    }
+    return ManagementFactory.getPlatformMBeanServer();
+  }
+
+  public void start() {
+    try {
+      // get all mbeans matching the query, register meters for them
+      Set<ObjectName> existingObjects = mBeanServer.queryNames(query.toObjectNameQuery(), null);
+      for (ObjectName existingObject : existingObjects) {
+        meters.compute(existingObject, this::computeMeters);
+      }
+
+      // register a notification listener that'll add and remove meters as they come
+      NotificationFilter filter = this::isNotificationEnabled;
+      notificationListener = this::handleNotification;
+      mBeanServer.addNotificationListener(
+          MBeanServerDelegate.DELEGATE_NAME, notificationListener, filter, null);
+    } catch (JMException e) {
+      log.warn("Could not start JMX metrics listener", e);
+    }
+  }
+
+  public void stop() {
+    if (notificationListener != null) {
+      try {
+        mBeanServer.removeNotificationListener(
+            MBeanServerDelegate.DELEGATE_NAME, notificationListener);
+        notificationListener = null;
+      } catch (InstanceNotFoundException | ListenerNotFoundException e) {
+        log.warn("Could not remove JMX metrics listener from the MBeanServer", e);
+      }
+    }
+
+    executorService.shutdown();
+
+    meters.values().stream().flatMap(Collection::stream).forEach(meterRegistry::remove);
+    meters.clear();
+  }
+
+  // visible for tests
+  Set<Meter> getMeters() {
+    return meters.values().stream().flatMap(Collection::stream).collect(toSet());
```

Comment:
most of this is duplicated with the above. You could factor out `getAllMeters()` or something that just encapsulates `meters.values().stream().flatMap(Collection::stream)`

Reply:
Good point, will do that. 👍